### PR TITLE
New GeoHexGrid aggregation

### PR DESCRIPTION
**Pull request by iverase**
This commit introduces a new geogrid aggregation called GeoHexGridAggregation that
is based in Uber h3 grid. It only supports geo_point fields.

### Docs preview
https://elasticsearch_82924.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/search-aggregations-bucket-geohexgrid-aggregation.html

### CODE REVIEW

**Comment by iverase** [comment]
@elasticmachine update branch

File: docs/reference/aggregations/bucket/geohexgrid-aggregation.asciidoc [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+==== Options
+
+[horizontal]
+field::         Mandatory. The name of the field indexed with GeoPoints.
+
+precision::     Optional. The integer zoom of the key used to define
+                cells/buckets in the results. Defaults to 6.
+                Values outside of [0,15] will be rejected.
+
+bounds::         Optional. The bounding box to filter the points in the bucket.
```
Comment (jrodewig) [inline docs/reference/aggregations/bucket/geohexgrid-aggregation.asciidoc:235]:
```suggestion
bounds::
(Optional, object) Bounding box used to filter the geo-points in each bucket.
Accepts the same bounding box formats as the
<<query-dsl-geo-bounding-box-query-accepted-formats,geo-bounding box query>>.
```

File: docs/reference/aggregations/bucket/geohexgrid-aggregation.asciidoc [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+size::          Optional. The maximum number of geohex buckets to return
+                (defaults to 10,000). When results are trimmed, buckets are
+                prioritised based on the volumes of documents they contain.
```
Comment (jrodewig) [inline docs/reference/aggregations/bucket/geohexgrid-aggregation.asciidoc:239]:
```suggestion
size::
(Optional, integer) Maximum number of buckets to return. Defaults to 10,000.
When results are trimmed, buckets are prioritized based on the volume of
documents they contain.
```

File: docs/reference/aggregations/bucket/geohexgrid-aggregation.asciidoc [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+shard_size::    Optional. To allow for more accurate counting of the top cells
+                returned in the final result the aggregation defaults to
+                returning `max(10,(size x number-of-shards))` buckets from each
+                shard. If this heuristic is undesirable, the number considered
+                from each shard can be over-ridden using this parameter.
```
Comment (jrodewig) [inline docs/reference/aggregations/bucket/geohexgrid-aggregation.asciidoc:245]:
```suggestion
shard_size::
(Optional, integer) Number of buckets returned from each shard. Defaults to
`max(10,(size x number-of-shards))` to allow for more a accurate count of the
top cells in the final result.
```

File: docs/reference/aggregations/bucket/geohexgrid-aggregation.asciidoc [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+          "key": "841fa4dffffffff",
+          "doc_count": 1
+        }
+      ]
+    }
+  }
+}
+--------------------------------------------------
+// TESTRESPONSE[s/\.\.\./"took": $body.took,"_shards": $body._shards,"hits":$body.hits,"timed_out":false,/]
+
```
Comment (jrodewig) [inline docs/reference/aggregations/bucket/geohexgrid-aggregation.asciidoc:104]: I'd insert the parameter definitions (currently the `Options` section) here. I'd also add a section immediately after with the response body definitions.

File: docs/reference/aggregations/bucket/geohexgrid-aggregation.asciidoc [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+==== Requests with additional bounding box filtering
+
+The `geohex_grid` aggregation supports an optional `bounds` parameter
+that restricts the cells considered to those that intersects the
+bounds provided. The `bounds` parameter accepts the bounding box in
```
Comment (jrodewig) [inline docs/reference/aggregations/bucket/geohexgrid-aggregation.asciidoc:173]:
```suggestion
provided bounds. The `bounds` parameter accepts the bounding box in
```

**Review by jrodewig — CHANGES_REQUESTED** [review]
Thanks @iverase! I left some suggestions and comments. I'd like to take another look, mostly to ensure I didn't introduce any errors in my review.

File: docs/reference/aggregations/bucket/geohexgrid-aggregation.asciidoc [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+The resulting grid can be sparse and only
+contains cells that have matching data. Each cell corresponds to a
+https://h3geo.org/docs/core-library/h3Indexing#h3-cell-indexp[H3 cell index] a
+and labeled using the https://h3geo.org/docs/core-library/h3Indexing#h3index-representation
+[H3Index representation].
+
+* High precision keys have a larger range for x and y, and represent tiles that
+cover only a small area.
+* Low precision keys have a smaller range for x and y, and represent tiles that
+each cover a large area.
```
[the remark this answers is not in the document]
Comment (iverase) [reply docs/reference/aggregations/bucket/geohexgrid-aggregation.asciidoc]: copy and paste bug ... I removed all together as you cannot really infer information easily from the key.

File: docs/reference/aggregations/bucket/geohexgrid-aggregation.asciidoc (hunk 4 above)
Comment (iverase) [reply docs/reference/aggregations/bucket/geohexgrid-aggregation.asciidoc:96]: Not sure what you are asking here. This is in line with what we have for other grid aggregations so if we change it here we should change for all of them.

**Review by jrodewig — APPROVED** [review]
Thanks @iverase.

I pushed a few small commits containing docs edits:
- https://github.com/elastic/elasticsearch/pull/82924/commits/1eebba32d0c35f7167b566e8011ce0608f800b6c
- https://github.com/elastic/elasticsearch/pull/82924/commits/73f09bbbaf81f1ff4160f7e7bcbb2337543116c4

If that looks good to you, these docs LGTM.

**Review by nik9000 — APPROVED** [review]
LGTM. All the interesting stuff is in `GeoHexCellIdSource` and it works like the other similar things. @imotov should also have a look, but it seems right to me.

**Review by imotov — APPROVED** [review]
LGTM

**Comment by jmp601** [comment]
> 

Getting a 404 on the doc. Would love to take a look and prepare what we have on the horizon for GEO.

**Comment by jrodewig** [comment]
Hi @jmp601, this work will be part of our upcoming 8.1.0 release. The docs for that release are currently in our `master` branch docs:
https://www.elastic.co/guide/en/elasticsearch/reference/master/search-aggregations-bucket-geohexgrid-aggregation.html

**Pull request merged**